### DISCUSSION TITLE
Is the plan for Contiki 4.9 to program motes only with sudo?

### DISCUSSION THREAD

**Discussion by rahav** [contiki-ng/contiki-ng#2562] [Q&A] [answered]
I tried it so far just with Nordic nrf52840 DK and it seems that the programming process will not work without `sudo`. Is this the plan moving forward @alexstanoev @wbober?

@pjonsson I don't think this relates to the other `sudo` discussion but keeping you in the loop.

Here is what I am seeing:

# Without `sudo`:
```
make TARGET=nrf52840 hello-world.upload
  OBJCOPY   build/nrf52840/dk/hello-world.elf --> build/nrf52840/dk/hello-world.i16hex
  SREC_CAT  build/nrf52840/dk/hello-world.i16hex --> build/nrf52840/dk/hello-world.hex
nrfjprog -f nrf52  --sectorerase --program build/nrf52840/dk/hello-world.hex
[error] [ Client] - Encountered error -105: Command connect_to_emu_with_snr executed for 218 milliseconds with result -105
ERROR: Unable to connect to a debugger.
[error] [ Worker] - JLinkARM.dll Open returned error 'Cannot connect to J-Link.'
ERROR: The JLinkARM DLL timed out while communicating to the J-Link probe.
ERROR: If the error condition persists, run the same command again with
ERROR: argument -- log, contact Nordic Semiconductor and provide the generated
ERROR: log.log file to them.
NOTE: For additional output, try running again with logging enabled (--log).
NOTE: Any generated log error messages will be displayed.
make: *** [../../arch/cpu/nrf52840/Makefile.nrf52840:129: hello-world.upload] Error 35

```

# With `sudo`:
```
sudo make TARGET=nrf52840 hello-world.upload
nrfjprog -f nrf52  --sectorerase --program build/nrf52840/dk/hello-world.hex
[ #################### ]   1.397s | Erase file - Done erasing                                                          
[ #################### ]   0.345s | Program file - Done programming                                                    
WARNING: A programming operation has been performed without --verify.
WARNING: Programming can fail without error.
nrfjprog -f nrf52  --reset
Applying system reset.
Run.
```

# Work environment
Contiki container running on Ubuntu 22.04 LTS

**Comment by rahav** [comment]
Interesting. I am now getting the following problems on some code examples. E.g. `examples/dev/leds`:

```
make TARGET=nrf52840
  MKDIR     build/nrf52840/dk/obj/.deps
mkdir: cannot create directory ‘build/nrf52840’: Permission denied
make: *** [../../../Makefile.include:239: build/nrf52840/dk/obj/.deps] Error 1
```

but it would work with `sudo`. This was not my experience before Contiki 4.9.

**Reply by pjonsson** [reply]
Running `sudo make ..` compiled something as root in `build/`. Do a `sudo rm -rf build` in the directory and things should work again.

**Comment by pjonsson** [comment, marked as answer]
I'm guessing there's something with the permissions on the USB devices in Linux, `/etc/udev/rules.d/99-jlink.rules` basically makes everything world writable. I have no idea how that is tunneled through to docker though, I have the jlink package installed in the Linux that I run the container in.

Here's the last time this was discussed: https://github.com/contiki-ng/contiki-ng/discussions/1794

**Reply by rahav** [reply]
I will reply here for both of your responses because they are related. You are correct that the directories were created as root. The challenge is that Nordic motes seem to require root privileges, then over time people will have such directories all over the place. At which point `make distclean` will not work as well, unless it runs with `sudo`.

I do not think it is the permissions on the USB devices in Linux because, for example, `sudo` is not required to program tmote sky with the same LEDs example. 

The docs will need to change to `sudo make ...`. Certainly new Contiki users may not remember which platform requires `sudo` and which does not.

I may be wrong about this, and I will explore it some more. I have different motes from Nordic and maybe it is Ubuntu version that is causing the problem. BTW @pjonsson, which version do you run under?

**Reply by alexstanoev** [reply]
Adding to what Andreas and Peter said, there's no requirement to be root to use nrfjprog (or any other Contiki-NG platform). 
The required permissions on the host are different to devices using a serial bootloader (i.e. Sky) - hence the aforementioned `/etc/udev/rules.d/99-jlink.rules` file.

Can you confirm your container is privileged and `/dev/bus/usb` is passed through? You will get this if you're using the contiker alias.

**Reply by rahav** [reply]
Yes, the container is invoked with the alias. Here is what I see inside the container
```
user@457b76dcb735:~/contiki-ng$ ls -l /dev/bus/usb/
total 0
drwxr-xr-x 2 root root 120 Jul 28 02:52 001
drwxr-xr-x 2 root root  60 Jul 28 02:52 002
```

The salient point is that without root, all `nrfjprog` commands fail. See my original post above please. It shows what occurs when it runs with and without `sudo`.

**Reply by pjonsson** [reply]
If you want to install lsusb, it's in the usbutils package, so `sudo apt update && sudo apt install usbutils` to install it in the container.

Here's what my machine says:
```
$ lsusb
Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 001 Device 028: ID 1366:1051 SEGGER J-Link
Bus 001 Device 027: ID 80ee:0021 VirtualBox USB Tablet
Bus 001 Device 026: ID 1915:520f Nordic Semiconductor ASA nRF52 USB CDC on Contiki-NG
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
$ ls -l /dev/bus/usb/001/
total 0
crw-rw-r-- 1 root root 189,  0 jul 22 12:44 001
crw-rw-rw- 1 root root 189, 25 jul 28 01:30 026
crw-rw-r-- 1 root root 189, 26 jul 28 01:30 027
crw-rw-rw- 1 root root 189, 27 jul 28 11:17 028
```
Device 026 is an nRF52840 dongle, and device 028 is what shows up when I turn on the nRF5340. The dongle required adding some extra line to the udev rules, but it still accomplishes the same thing, making the devices world writable.

As I mentioned, I have the jlink package installed outside the container too, so it's possible that is what triggers and makes my devices world writable. I use the contiker alias, but I don't run anything explicitly as root.

**Reply by rahav** [reply]
Thank you all. My system does not work yet, but making progress. Please see below, following @pjonsson post.

When the nRF52840-DK is connected and ON. In my system I have:
```
lsusb
Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 001 Device 002: ID 413c:301a Dell Computer Corp. Dell MS116 Optical Mouse
Bus 001 Device 006: ID 1366:1051 SEGGER J-Link
Bus 001 Device 004: ID 0bda:c829 Realtek Semiconductor Corp. Bluetooth Radio 
Bus 001 Device 003: ID 413c:2113 Dell Computer Corp. KB216 Wired Keyboard
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
```

So following your lead, the mote is device 006 on BUS 001 and the permissions on this device on my system are:
```
ls -l /dev/bus/usb/001
total 0
crw-rw-r-- 1 root root 189, 0 Jul 28 02:52 001
crw-rw-r-- 1 root root 189, 1 Jul 28 02:52 002
crw-rw-r-- 1 root root 189, 2 Jul 28 02:52 003
crw-rw-r-- 1 root root 189, 3 Jul 28 02:52 004
crw-rw-r-- 1 root root 189, 5 Jul 28 12:02 006
```
**So the Nordic device is NOT world writable on my system**. This is a new computer and fresh installation of Contiki 4.9. Any idea what went wrong on my system? 

Which one of the following looks promising for me to explore?
1. What Ubuntu version are you running? I am on Ubuntu 22.04 LTS, should I install a different OS?
2. I can download a new Contiki 4.9 container, but maybe something is wrong with my host file system.
3. Are we suppose to edit the udev rules file?
4. So far I worked with the Nordic DK but I also have the Nordic dongle. I haven't used it yet. Did you need to modify the udev rules file to make it work?

**Reply by pjonsson** [reply]
Your versions and downloads are fine.

I don't know how to make Docker do what you want, but hopefully Alex can answer that. If you don't mind installing the JLink package on the host, my guess is that things would work for you. Here's what the current Docker file uses:

```shell
wget -nv https://nsscprodmedia.blob.core.windows.net/prod/software-and-other-downloads/desktop-software/nrf-command-line-tools/sw/versions-10-x-x/10-22-0/nrf-command-line-tools_10.22.0_amd64.deb
sudo apt-get -y --no-install-recommends install ./nrf-command-line-tools_10.22.0_amd64.deb
sudo apt-get -y --no-install-recommends install /opt/nrf-command-line-tools/shareJLink_Linux_V780c_x86_64.deb
```
The dongle requires an extra udev rule, the line is described in the Contiki-NG documentation for that platform. I put that line first in the jlink rules since I don't understand the flow of initialization for udev, but I'm fairly sure that line can go in any other file as well, as long as it gets executed.

**Reply by rahav** [reply]
You are a very smart man @pjonsson. It works.

I believe that every user will run into this problem with Nordic devices. It probably did not come up in testing because testers with lots of Nordic hours coincidentally already had the nRF Command Line Tools installed on the host machine. How do we change the docs to tell people to install nRF Command Line Tools and SEGGER J-Link (I had to install both) on the host machine? I think we need to do that unless there is a way to fix the container.

# Install Nordic nRF Command Line Tools and SEGGER J-Link on a host computer
BTW, I couldn't install using wget (see appendix below). So here are the instructions for Ubuntu:

On the host computer (not inside the Contiki container), download the version for your host OS from [Nordic nRF Command Line Tools](https://www.nordicsemi.com/Products/Development-tools/nRF-Command-Line-Tools/Download?lang=en#infotabs)

Install it
```
sudo dpkg -i Downloads/nrf-command-line-tools_10.23.0_amd64.deb
```

Install SEGGER J-Link
```
sudo apt install /opt/nrf-command-line-tools/share/JLink_Linux_V788j_x86_64.deb --fix-broken
```



# Appendix
1. We may be able to skip the download and directly do `apt install` for the nRF tools as well.

2. The wget command, copied from the Contiki container, fails to install.
```
wget -nv https://nsscprodmedia.blob.core.windows.net/prod/software-and-other
-downloads/desktop-software/nrf-command-line-tools/sw/versions-10-x-x/10-22-0/nr
f-command-line-tools_10.22.0_amd64.deb && \
sudo apt-get -y --no-install-recommends install ./nrf-command-line-tools_10.22.
0_amd64.deb && \
sudo apt-get -y --no-install-recommends install /opt/nrf-command-line-tools/share/JLink_Linux_V780c_x86_64.deb
https://nsscprodmedia.blob.core.windows.net/prod/software-and-other:
2023-07-28 11:19:10 ERROR 404: The specified blob does not exist..
bash: -downloads/desktop-software/nrf-command-line-tools/sw/versions-10-x-x/10-22-0/nr: No such file or directory
f-command-line-tools_10.22.0_amd64.deb: command not found
0_amd64.deb: command not found
```

**Reply by pjonsson** [reply]
My cut and paste from the Dockerfile had broken line breaks, fixed the comment now.

**Comment by arurke** [comment]
@rahav Are you starting the contiki-ng container using the [alias from the doc.](https://docs.contiki-ng.org/en/develop/doc/getting-started/Docker.html), including the `--privileged` flag?
```
export CNG_PATH=<absolute-path-to-your-contiki-ng>
alias contiker="docker run --privileged --sysctl net.ipv6.conf.all.disable_ipv6=0 --mount type=bind,source=$CNG_PATH,destination=/home/user/contiki-ng -e DISPLAY=$DISPLAY -e LOCAL_UID=$(id -u $USER) -e LOCAL_GID=$(id -g $USER) -v /tmp/.X11-unix:/tmp/.X11-unix -v /dev/bus/usb:/dev/bus/usb -ti contiker/contiki-ng"
```

**Reply by rahav** [reply]
Yes, these two lines are in my `.bashrc`, and here is how the alias is defined:
```
type contiker
contiker is aliased to `docker run --privileged --sysctl net.ipv6.conf.all.disable_ipv6=0 --mount type=bind,source=/home/rahav/contiki-ng,destination=/home/user/contiki-ng -e DISPLAY=:0 -e LOCAL_UID=1000 -e LOCAL_GID=1000 -v /tmp/.X11-unix:/tmp/.X11-unix -v /dev/bus/usb:/dev/bus/usb -ti contiker/contiki-ng'
```
